distributed/deploy/adaptive.py

```diff
@@ -123,17 +123,32 @@ def needs_cpu(self):
         Notes
         -----
         Returns ``True`` if the occupancy per core is some factor larger
-        than ``startup_cost``.
+        than ``startup_cost`` and the number of tasks exceeds the number of
+        workers
         """
         total_occupancy = self.scheduler.total_occupancy
         total_cores = sum([ws.ncores for ws in self.scheduler.workers.values()])
 
         if total_occupancy / (total_cores + 1e-9) > self.startup_cost * 2:
             logger.info("CPU limit exceeded [%d occupancy / %d cores]",
                         total_occupancy, total_cores)
-            return True
-        else:
-            return False
+
+            num_workers = len(self.scheduler.workers)
```

Comment:
Shpuldn't you use total_cores for the number of tasks comparison? Do you have one thread per worker in your setup?

Reply:
hmm - I'm not sure I follow, but I am on a dask_kubernetes setup with 1:1 threads to workers, so maybe my use case isn't universal. I think I see, though... in many cases the number of dask processes/threads which can take on tasks is greater than the number of workers? Is total_cores the number of workers * (threads per worker or procs per worker)?

Reply:
> I am on a dask_kubernetes setup with 1:1 threads to workers, so maybe my use case isn't universal
> in many cases the number of dask processes/threads which can take on tasks is greater than the number of workers

Yes!

I think in most cases having several threads per worker is better, except for GIL boud tasks. Dask-kubernetes propose 2 threads by default, which is low but it depends on VMs you use:

https://github.com/dask/dask-kubernetes/blob/master/dask_kubernetes/kubernetes.yaml#L24-L26

With dask-jobqueue, I often use 4, 8 or even 24 cores per worker.

> Is total_cores the number of workers * (threads per worker)?

Yes!

Reply:
Ok thanks! Just so I understand before modifying the code again... should I take this to mean that dask uses the term "core" to refer specifically to the number of tasks that can be handled simultaneously by a worker, regardless of the hardware? Or are you doing something clever with the relationship between physical cores and CPU needs that I'm not following? Thanks again for the helping hand in this :)

Reply:
ok this makes a lot of sense. so I should count the number of available threads _or_ processes, not the number of workers

Reply:
A Worker = 1 process = several threads, on scheduler side. So just using the already computed `total_cores` should be fine.